modules/tinymce/src/themes/silver/test/ts/headless/components/tree/TreeTest.ts

```diff
@@ -0,0 +1,144 @@
+              type: 'menubutton',
+              fetch: (success) => success([
+                {
+                  type: 'menuitem',
+                  text: 'menuitem',
+                  onAction: () => {
+                    store.add('menuitem');
+                  }
+                }
+              ])
+            }
+          },
+        ]
+      },
+      {
+        type: 'leaf',
+        title: 'File 5',
+        id: '5',
+      },
+      {
+        type: 'leaf',
+        title: 'File 6',
+        id: '6',
+      },
+    ];
+
+    const treeSpec = StructureSchema.getOrDie(Dialog.createTree({
+      type: 'tree',
+      onLeafAction: store.add,
+      items: fullTree
+    }));
+
+    const tree = renderTree(treeSpec, extrasHook.access().extras.backstages.dialog );
+    return GuiFactory.build({
+      dom: {
+        tag: 'div'
+      },
+      components: [ tree ]
+    });
+  });
+
+  const assertDirectoryExpandedState = (label: string, expected: boolean, directory: AlloyComponent) => {
+    assert.equal(Class.has(directory.element, 'tox-trbtn--expanded'), expected, 'Checking if expanded class is present: ' + label);
+  };
+
+  const getTreeItem = (selector: string) => {
+    const component = hook.component();
+    return component.getSystem().getByDom(
+      SelectorFind.descendant(component.element, selector).getOrDie(
+        `Could not find tree item defined by: ${selector}`
+      )
+    ).getOrDie();
+  };
+
+  it('Check initial event state', () => {
+    const store = hook.store();
+    store.assertEq('Store should empty', []);
+  });
+
+  it('Tree', async () => {
+    const dir = getTreeItem('.tox-tree--directory');
+    const store = hook.store();
+    store.clear();
+    Assertions.assertStructure(
+      'Checking directory structure',
+      ApproxStructure.build((s, _str, arr) => s.element('div', {
+        classes: [ arr.has('tox-tree--directory') ],
+      })),
+      dir.element
+    );
+
+    assertDirectoryExpandedState('Collapsed', false, dir);
+    Mouse.clickOn(dir.element, '.tox-trbtn.tox-tree--directory__label');
+    assertDirectoryExpandedState('Expanded', true, getTreeItem('.tox-tree--directory__children'));
+
+    assertDirectoryExpandedState('Collapsed', false, getTreeItem('.tox-tree--directory .tox-tree--directory .tox-tree--directory__children'));
+    Mouse.clickOn(dir.element, '.tox-tree--directory .tox-trbtn.tox-tree--directory__label');
+    assertDirectoryExpandedState('Expanded', true, getTreeItem('.tox-tree--directory .tox-tree--directory .tox-tree--directory__children'));
+
+    Mouse.clickOn(getTreeItem('.tox-tree').element, '>.tox-tree--leaf__label');
+    store.assertEq('File 5', [ '5' ]);
+
+    store.clear();
+    Mouse.clickOn(getTreeItem('.tox-tree').element, '.tox-tree--leaf__label');
+    store.assertEq('File 1', [ '1' ]);
+
+    store.clear();
+    Mouse.clickOn(dir.element, '.tox-mbtn');
+    // Wait for the menu item to show up
+    await Waiter.pWait(1);
```

Comment:
Can this be a predicate waiter or selector waiter so we don't get some flake here?

Reply:
Yes, I changed this to use `UiFinder.pWaitFor`.